Add Ondilo ico integration

JeromeHXP opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Adds a new component to be able to retrieve Ondilo ICO data using Ondilo's APIs. ICOs are small connected devices that you simply drop into the pool or spa and continuously analyzes the water quality.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16055

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Commit:
First implementationof Ondilo component support

Commit:
Update manifest toadd pypi pkg dependency

Commit:
Update entities name and corrected refresh issue

Commit:
Changed percentage unit name

Commit:
Corrected merge issues

Commit:
Updated coveragerc

Commit:
cleaned up code and corrected config flow tests

homeassistant commented:
Hi @JeromeHXP, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#44728).

Once you do that we will be able to review and accept this pull request.

Thanks!

Commit:
Code cleanup and added test for exisitng entry

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for devidx, _ in enumerate(pool["devices"]):
```
Optional: Consider if its worth being defensive about ensuring "data_type" is found in SENSOR_TYPES

allenporter replied:
Alternatively: iterate over every supported SENSOR_TYPE and check for it as a measure that exists in the set of measures, then create the entities just for those.  (This would only work if you did the "id lookup" pattern below and made the sensor data indexed by measure type)

JeromeHXP replied:
Done, thanks.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for devidx, _ in enumerate(pool["devices"]):
+            entities.append(OndiloICO(coordinator, poolidx, devidx))
```
I see this is following the pattern in the docs, however I was curious what would happen if the API started returning devices in a different order if a new device was created and added to the front of the list.  Is this a non-issue? I realize this is beyond your component.  I was imagining some kind of "id lookup" would be more resilient to problems than an index lookup, but not sure how real of a problem that is.

JeromeHXP replied:
Interesting point thanks. As a JSON object is unordered, this could indeed happen, although I doubt Ondilo would change that order. Anyway, I've taken that into account and changed the behavior to look for the "data_type" field. And I did the same for the pools and relied on the Ondilo pool Id.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+            for pool in pools:
+                pool["ICO"] = await hass.async_add_executor_job(
```
Consider if it makes sense to have a single method blocking method that does these two synchronous calls, then have an `await asyncio.gather` at the end to join all the pools together at the end.  Something kind of like what https://github.com/home-assistant/core/pull/44376/files did

JeromeHXP replied:
Interesting point. Done.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+        self._devidx = devidx
+        self._poolidx = poolidx
+        self._unique_id = (
+            f"{self._pooldata()['ICO']['serial_number']}-{self._devdata()['data_type']}"
```
Did you consider using `uuid` here? Curious how you think about that compared to serial_number.

JeromeHXP replied:
Well, this couple uniquely identifies a sensor, so I thought I could use that and that it could be easier to troubleshoot. Not so sure about that any more but it doesn't harm I think. Would you rather use a UUID? If that's usually what's recommended, that's not hard to change (although might be an issue for compatibility with those already using this component as a custom_component)

allenporter replied:
You'd know best, I was just making observations reading the API docs.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+                pool["ICO"] = await hass.async_add_executor_job(
+                    api.get_ICO_details, pool["id"]
+                )
+                pool["devices"] = await hass.async_add_executor_job(
```
Terminology nit: It feels like "ICO" is the device, and this is a list of "sensors" rather than "devices".  What do you think?

JeromeHXP replied:
Correct thanks.

JeromeHXP committed:
Changes following PR comments:
- Inherit CoordinatorEntity instead of Entity
- Merged pools blocking calls into one
- Renamed devices vars to sensors
- Check supported sensor types
- Stop relying on array index position for pools
- Stop relying on attribute position in dict for sensors

JeromeHXP committed:
Corrected unit test

JeromeHXP commented on homeassistant/components/ondilo_ico/__init__.py:
```diff
@@ -0,0 +1,59 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Ondilo ICO from a config entry."""
+
+    config_flow.OAuth2FlowHandler.async_register_implementation(
```
As far as I can tell, when using the config flow, async_setup or async_setup_entry are not called before the call to async_step_user, so I need that call in async_step_user. Then, when restarting hass, I also need to register the OAuthImplementation so I need to make sure it's declared either in async_setup or in async_setup_entry. As I don't support the yaml declaration, I chose to put that call in async_setup_entry. Defining it only in async_setup is not enough.
Am I wrong here?

allenporter replied:
Yeah, that makes sense.  I see what I was missing.  I got this from https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow_oauth2/integration/__init__.py#L42  The difference may just be that in that example a `configuration.yaml` is needed for configuring oauth client id and client secret, and that would invoke `async_setup`. Not the case here since its hard coded.

allenporter replied:
Do you know if you actually need to call `async_register_implementation` or `async_get_config_entry_implentation` here? It may be sufficient to just pass `OndiloOauth2Implementaiton` into `OndiloClient` since the config flow will register itself.

JeromeHXP commented on homeassistant/components/ondilo_ico/translations/en.json:
```diff
@@ -0,0 +1,17 @@
+{
```
It was created using `python3 -m script.translations develop`. Should I remove it? Worst case, I guess, it will just be recreated afterwards?

JeromeHXP commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,186 @@
+    async_add_entities(entities)
+
+
+class OndiloICO(Entity):
```
Thanks, I can see this was indeed updated in the doc (it wasn't there when I first did that in August). I'll update accordingly.

allenporter commented:
I'm happy to look again -- The diffs look like they may be including other stuff, so maybe a rebase is in order?

JeromeHXP commented:
> I'm happy to look again -- The diffs look like they may be including other stuff, so maybe a rebase is in order?

@allenporter Yep sorry I messed up with my rebase. It should be clean now.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,185 @@
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for sensor_idx, sensor in enumerate(pool["sensors"]):
+            if SENSOR_TYPES.get(sensor["data_type"], None) is not None:
```
I believe the `None` argument to `get` is the default and can be omitted.  Also the "is not None" can be omitted. Another option is:

```
if sensor["data_type"] in SENSOR_TYPES:
```

JeromeHXP replied:
Exact... Not sure why I chose this way.... Updated.

allenporter commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,185 @@
+    def _pooldata(self):
+        """Get pool data dict."""
+        return next(
+            (pool for pool in self.coordinator.data if pool["id"] == self._poolid),
```
I could see doing this work up front in `async_update_data` as another approach to consider in a future PR, though i realize the N count is probably low here.

JeromeHXP replied:
Correct. Still, since I was also worried of that, I try to avoid multiple calls to it (and `_devdata`) in the same function. That being said, you're right, I think most people only have one ICO so it shouldn't be an issue.

JeromeHXP committed:
Reformat sensor type check

allenporter commented:
Looks great, happy new year!

frenck commented on homeassistant/components/ondilo_ico/strings.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "title": "Ondilo ICO",
```
I don't think this title needs to be translated?

JeromeHXP replied:
No it doesn't. Does it mean I should remove it from the file?

frenck replied:
Yes, because this adds it to translations 😉

JeromeHXP replied:
OK Done and pushed to my branch. Should I open a new PR as this one is closed?

allenporter replied:
Yeah, and I'm happy to get tagged to review quickly

JeromeHXP replied:
@allenporter Done. PR is here: https://github.com/home-assistant/core/pull/44824

MartinHjelmare commented on homeassistant/components/ondilo_ico/manifest.json:
```diff
@@ -0,0 +1,18 @@
+  "requirements": [
+    "ondilo==0.2.0"
+  ],
+  "ssdp": [],
```
Please remove empty items.

JeromeHXP replied:
Done in PR https://github.com/home-assistant/core/pull/44837

MartinHjelmare commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,185 @@
+    "temperature": [
+        "Temperature",
+        TEMP_CELSIUS,
+        "mdi:thermometer",
```
We don't set a custom icon when we set a device class. Those will already have specific icons in the frontend.

JeromeHXP replied:
Done in PR https://github.com/home-assistant/core/pull/44837

MartinHjelmare commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,185 @@
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+
+            return await asyncio.gather(
```
We don't use asyncio.gather with the executor to not greedily consume many threads from the pool. It's best to just run all sync calls in a single function that we schedule on the executor to minimize context switches.

JeromeHXP replied:
Done in PR https://github.com/home-assistant/core/pull/44837

MartinHjelmare commented on homeassistant/components/ondilo_ico/sensor.py:
```diff
@@ -0,0 +1,185 @@
+    @property
+    def state(self):
+        """Last value of the sensor."""
+        _LOGGER.debug(
```
We don't do side effects, like logging, in entity state properties. Please remove the logging.

JeromeHXP replied:
Done in PR https://github.com/home-assistant/core/pull/44837

MartinHjelmare commented on tests/components/ondilo_ico/test_config_flow.py:
```diff
@@ -0,0 +1,88 @@
+    """Check flow abort when an entry already exist."""
+    MockConfigEntry(domain=DOMAIN).add_to_hass(hass)
+
+    flow = config_flow.OAuth2FlowHandler()
```
Don't create the flow directly. Start the test with `hass.config_entries.flow.async_init`.

JeromeHXP replied:
Done in PR https://github.com/home-assistant/core/pull/44837

MartinHjelmare reviewed (commented):
Nice! Please address the comments in a future PR.
